`FusedAdam(W)` should take `OptState` into account before unscaling grads

crcrpar opened the pull request:
the optimizers have to consult `OptState` before unscaling gradients because we could call `GradScaler.unscale_` explicitly to for e.g. `clip_grad_norm_` as mentioned in https://github.com/pytorch/pytorch/blob/e52786f3d177a7ca5d490a516cf52e236ef072cb/torch/cuda/amp/grad_scaler.py#L235-L266 and https://pytorch.org/docs/stable/notes/amp_examples.html#working-with-unscaled-gradients

Related #90752

albanD commented on torch/cuda/amp/grad_scaler.py:
```diff
@@ -355,11 +355,13 @@ def step(self, optimizer, *args, **kwargs):
                 )
                 optimizer.grad_scale = scaler
                 optimizer.found_inf = found_inf
+                optimizer.grads_unscaled = optimizer_state["stage"] == OptState.UNSCALED
```
Is there any case where we want to pass a non-None `grad_scale` and `grads_unscale == True` ? If so, is the new flag actually redundant and we could just modify above:
```python
optimizer.grad_scale = scaler if optimizer_state["stage"] == OptState.UNSCALED else None
```

crcrpar replied:
oh it looks right. thank you for the swift response

crcrpar committed:
Set `grad_scale` after consulting `OptState`

We would call `GradScaler.unscale_()` explicitly before
`GradScaler.update()` when we want to e.g.  clip gradients.
Thus we have to take `OptState` into consideration and skip gradient
unscaling if OptState is OptState.UNSCALED.

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

albanD reviewed (approved):
That is a lot simpler :D 
Thanks for the fix!

crcrpar commented:
@pytorchbot merge
